### PR TITLE
refactor(compiler-cli): Enable pipe information when checkTypeOfPipes…

### PR DESCRIPTION
**Pull request by atscott**
…=false

When `checkTypeOfPipes` is set to `false`, the configuration is meant to
ignore the signature of the pipe's `transform` method for diagnostics.
However, we still should produce some information about the pipe for the
`TemplateTypeChecker`. This change refactors the returned symbol for
pipes so that it also includes information about the pipe's class
instance as it appears in the TCB.

### CODE REVIEW

**Comment by atscott** [comment]
Adding blocked label and converting to draft - this might want to wait on the changes from @alxhub to the LS testing infrastructure so we can add tests to the LS with different compiler options.

File: packages/compiler-cli/src/ngtsc/typecheck/api/symbols.ts [hunk 1]
```diff
@@ -247,3 +248,37 @@ export interface DomBindingSymbol {
...
+
+  /** The position of the transform call in the template. */
+  shimLocation: ShimLocation;
+
+  /** The symbol for the pipe class as an instance that appears in the TCB. */
+  classSymbol: PipeClassSymbol;
+}
+
+/** Represents an instance of a pipe found in the TCB, i.e. `var _pipe1: MyPipe = null!; */
+export interface PipeClassSymbol {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/api/symbols.ts:275]: How about just `ClassSymbol`, as this actual interface definition has little to do with pipes.

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts [hunk 2]
```diff
@@ -354,6 +356,45 @@ export class SymbolBuilder {
     }
   }
 
+  private getSymbolOfPipe(expression: BindingPipe): PipeSymbol|null {
+    const node = findFirstMatchingNode(
+        this.typeCheckBlock, {withSpan: expression.sourceSpan, filter: ts.isCallExpression});
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:361]: Should we validate that the actual method being called is named `transform` here? Or is that overkill?

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts (hunk 2 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/typecheck/src/template_symbol_builder.ts:361]: I think that's a bit overkill but I won't fight it if you think we should. My opinion is that this falls in the realm of trusting our spans in the TCB to be on the correct nodes. I don't know if there's a world where we would put the expression's source span on another call expression.

File: packages/language-service/ivy/test/definitions_spec.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        export class AppModule {}
+      `,
+        isRoot: true
+      },
+      {
+        name: absoluteFrom('/app.html'),
+        contents: `Will be overridden`,
+      }
+    ];
+    env = LanguageServiceTestEnvironment.setup(testFiles, {strictTemplates: false});
```
Comment (kyliau) [inline packages/language-service/ivy/test/definitions_spec.ts:40]: Could you please add the same comment here?
>// checkTypeOfPipes is set to false when strict templates is false

File: packages/language-service/ivy/test/test_utils.ts [hunk 4]
```diff
@@ -51,3 +53,25 @@ export function createModuleWithDeclarations(
...
+
+export interface HumanizedDefinitionInfo {
+  fileName: string;
+  textSpan: string;
+  contextSpan: string|undefined;
+}
+
+export function humanizeDefinitionInfo(
+    def: ts.DefinitionInfo, host: MockServerHost,
+    overrides: {[fileName: string]: string} = {}): HumanizedDefinitionInfo {
```
Comment (kyliau) [inline packages/language-service/ivy/test/test_utils.ts:65]: nit: consider using a Map here, which provides better semantics and typings.

File: packages/language-service/ivy/test/type_definitions_spec.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        export class AppModule {}
+      `,
+        isRoot: true
+      },
+      {
+        name: absoluteFrom('/app.html'),
+        contents: `Will be overridden`,
+      }
+    ];
+    env = LanguageServiceTestEnvironment.setup(testFiles, {strictTemplates: false});
```
Comment (kyliau) [inline packages/language-service/ivy/test/type_definitions_spec.ts:40]: Please add comment
>// checkTypeOfPipes is set to false when strict templates is false

**Review by kyliau — APPROVED** [review]

**Pull request closed without merging**